Update the testing matrix (5 distros, 2 python versions, 3 salt versions)

javierbertoli opened the pull request:
As discussed in the [Slack formula's channel](https://saltstackcommunity.slack.com/messages/C7LG8SV54/), tried to cover most cases, so I ended with this matrix:

#### salt 2019.2, python 3 (installed from Saltstack's repo)
* debian-9
* ubuntu-18.04
* centos-7

#### salt 2019.2, python 2 (installed from Fedora's repo)
* fedora-29

#### salt 2018.3, python 2 (installed from Saltstack's repo)
* opensuse-42.3
* debian-8
* ubuntu-16.04
* centos-6

#### salt 2018.3, python 2 (installed from Fedora's repo)
* forced-fedora-28: need to force the version, otherwise, `dnf` installs 2019.2 from the `updates` repo

#### salt 2017.7, python 2 (installed from Saltstack's repo)
* debian-8
* ubuntu-16.04
* centos-6

Added tests to verify the package version in all cases, and split the tests dirs for the three salt versions, so we can now test config correctly for them.

myii commented on Gemfile:
```diff
@@ -1,7 +1,6 @@
-gem "kitchen-salt", ">=0.2.5"
-gem "kitchen-inspec"
+gem 'kitchen-docker', '>= 2.9'
+gem 'kitchen-salt', '>= 0.5.0'
```
We can use `0.6.0` now.

myii commented on kitchen.yml:
```diff
@@ -0,0 +1,135 @@
+driver:
+  name: docker
+
+driver_config:
```
As @dafyddj mentioned in Slack, lines 4 and 5 here can be removed, so that we end up with:

```yaml
driver:
  name: docker
  use_sudo: false
  privileged: true
  provision_command: mkdir -p /run/sshd
  run_command: /lib/systemd/systemd
```

dafyddj replied:
The same is true for lower down under `platforms:` too.

myii commented on kitchen.yml:
```diff
@@ -0,0 +1,135 @@
+  # Previous distro, oldest salt, previous python
+  - name: debian-8-2017-7-py2
+    driver_config:
+      image: saltstack/au-debian-8:ci-2017.7-py2
```
`image: registry.gitlab.com/dafyddj/debian-salt:8-2017.7`

myii commented:
@javierbertoli There's a strange problem: most images appear to end up with a Salt version that doesn't match the image that is being pulled.  Based upon searching the log for `sweet!`.  For example:

* `salt-debian-9-2019-2-py3` reports as `2018.3.3`.
* `salt-ubuntu-1804-2019-2-py3` => `2018.3.3`.
* `salt-centos-7-2019-2-py3` => `2018.3.3`.
* Etc.

javierbertoli commented:
@myii interesting, it seems we have a few issues here:

1. The docker images don't ship what they say. On a freshly pulled image (expected salt 2019.2, python3):
```bash
$ docker run -ti saltstack/au-debian-9:ci-2019.2-py3 /bin/bash
root@be157c63728b:/# salt --version
salt 2018.3.3 (Oxygen)

root@be157c63728b:/# head -2 /usr/bin/salt
#!/usr/bin/python

root@be157c63728b:/# python --version
Python 2.7.13
```

2. [This line](https://github.com/saltstack-formulas/salt-formula/blob/master/.kitchen.yml#L29), which I was aware that needed to be changed on suite-by-suite basis, is failing (as you mention) because running salt's bootstrapping script on the container above effectively updates it to latest and it's not happening on the kitchen-ci runs. 
  I wonder if [these locale issues](https://travis-ci.com/saltstack-formulas/salt-formula/jobs/188205826#L1009) are causing the installer to fail. 
  I took a quick look at the [kitchen-salt installer file](https://github.com/saltstack/kitchen-salt/blob/master/lib/kitchen/provisioner/install.erb) and, comparing the lines on it with the output of the [debian run logs](https://api.travis-ci.com/v3/job/188205826/log.txt) it seems that the installer script is not being executed.

myii commented:
@daks @aboe76 This is a fantastic effort from @javierbertoli.  It would be great to get feedback from both of you, so that we can confirm this is how we want to proceed for this and other formulas as well.

@javierbertoli As mentioned on Slack, just one slight discrepancy in the log:

* `v2019-2-py3-centos-7` is showing as Python 2.

myii commented:
@javierbertoli One request: we've run into bugs with `salt-ssh`, such as [this one](https://github.com/saltstack/salt/issues/51605).  Would it be possible to have one of the instances be a `salt-ssh` instance?

daks commented on kitchen.yml:
```diff
@@ -0,0 +1,208 @@
+  - name: ubuntu-16.04
+  - name: fedora-28
+  # Fails on Debian due to vsyscall issues, see
+  # https://hub.docker.com/_/centos, "A note about vsyscall"
```
Are you talking about Debian or Centos?

javierbertoli replied:
Need to elaborate these comments better, atm they're my notes during the attempts to get this working: it means `centos-6` *guest* has issues on `debian` *hosts*, due to Debian having disabled `vsyscall`.

javierbertoli replied:
Re-worded the comment to make it a bit more clear.

daks commented:
Looks good to me, good job @javierbertoli :)

javierbertoli commented:
> @javierbertoli As mentioned on Slack, just one slight discrepancy in the log:
> 
>    v2019-2-py3-centos-7 is showing as Python 2.

Fixed. There's were three issues that caused this:

1. the salt's bootstrapper, at the begining of the kitchen run was running with default parameters and therefore installing `python2` version of salt packages
2. when adding the repo, [the repo name and file](https://github.com/saltstack-formulas/salt-formula/blob/4231029c6296d6ce44497cb15232ef25c1533cf8/salt/pkgrepo/redhat/init.sls#L4-L6) (saltstack-repo) were different than the ones installed by the bootstrapper (saltstack). Therefore, we ended up with two repos, one for `python3` (the desired one) and one for `python2`. The same was happenening before for suse, that's why I added [that weird repo name](https://github.com/saltstack-formulas/salt-formula/blob/4231029c6296d6ce44497cb15232ef25c1533cf8/salt/pkgrepo/suse/init.sls#L5) there.
3. And finally, when the `state` checked for the package, in the case of 2 (2 repos), it either found the package installed and did nothing (leaving a python2 packag) or, when fixing the repo, [found inconsistent yum metadata and failed](https://github.com/saltstack-formulas/salt-formula/blob/4231029c6296d6ce44497cb15232ef25c1533cf8/kitchen.yml#L88-L94).

The changes pushed should fix all that. Bear in mind that these are weird situations caused by the `kitchen-salt` bootstrapping of salt to test formulas, and a `salt formula` trying to install salt in an instance with salt already installed on the bootstrapper. Too many hands in one plate :yum: 

> @javierbertoli One request: we've run into bugs with salt-ssh, such as this one. Would it be possible to have one of the instances be a salt-ssh instance?

It's possible. Never tried `salt-ssh` in `kitchen`, but I think is doable. Any suggestion of what to test? Or else, we can submit another PR to address that?

myii commented:
@javierbertoli As discussed on Slack just now, we'll leave this `salt-ssh` for a future PR.  I'm happy with this PR to be merged, thanks again for the dedication!  The folks at Travis CI are probably breathing a collective sigh of relief that this PR testing is over!

aboe76 commented:
@javierbertoli and @myii what a work!!!, 

I can only see one issue coming up, soon, when you switch to opensuse leap 15, it will be python3...

javierbertoli commented:
There's an [opensuse/leap:15](https://hub.docker.com/r/opensuse/leap/tags) image available, I can switch to/add it to the matrix if you want. 

We discussed it with @myii on slack and, as none of us is an active opensuse user, we are not aware of which version to use/prefer. So I went with the 42.3 because I remember we used it somewhere in the formulas.

But if you have a preference/suggestion of which opensuse images to use, I think I can change it. Either in this PR or adding a new issues/PR for it.

javierbertoli commented:
@aboe76, tried to change the code to `opensuse Leap 15`, but hit an issue with `inspec`: [they're not currently supporting it](https://github.com/inspec/inspec/blob/master/README.md#supported-os), only 13 & 42 and therefore the tests are skipped

```
  ↺  salt packages: should be installed (2 skipped)
     ↺  The `package` resource is not supported on your OS yet.
     ↺  The `package` resource is not supported on your OS yet.
  ↺  salt services: should be running (2 skipped)
     ↺  The `service` resource is not supported on your OS yet.
     ↺  The `service` resource is not supported on your OS yet.
```
All in all, given that opensuse is in the matrix and is being tested, I think that we can upgrade to `opensuse-leap-15` in a couple of months, or submit a patch to `inspec`? wdyt?

aboe76 commented:
Thanks for trying, you did an amazing job.


Op di 2 apr. 2019 00:42 schreef Javier Bértoli <notifications@github.com>:

> @aboe76 <https://github.com/aboe76>, tried to change the code to opensuse
> Leap 15, but hit an issue with inspec: they're not currently supporting it
> <https://github.com/inspec/inspec/blob/master/README.md#supported-os>,
> only 13 & 42 and therefore the tests are skipped
>
>   ↺  salt packages: should be installed (2 skipped)
>
>      ↺  The `package` resource is not supported on your OS yet.
>
>      ↺  The `package` resource is not supported on your OS yet.
>
>   ↺  salt services: should be running (2 skipped)
>
>      ↺  The `service` resource is not supported on your OS yet.
>
>      ↺  The `service` resource is not supported on your OS yet.
>
>
> All in all, given that opensuse is in the matrix and is being tested, I
> think that we can upgrade to opensuse-leap-15 in a couple of months, or
> submit a patch to inspec? wdyt?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/saltstack-formulas/salt-formula/pull/406#issuecomment-478773621>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABt51MGJF_a8R02AXDvMwTLXSFzuUJ_Wks5vcotkgaJpZM4cNpBU>
> .
>

myii commented:
@javierbertoli Merged -- thanks for this excellent contribution.

To everyone else, your feedback and comments were extremely helpful and much appreciated.
